### PR TITLE
Add new Unmanaged MD calling convention value

### LINKED ISSUE #38135: [Proposal] `System.Runtime.CompilerServices.RuntimeFeature.UnmanagedCallKind`
## Background and Motivation

With .NET 5, we're going to be adding support for a new `CallKind` bit (tracked here: https://github.com/dotnet/runtime/issues/38133). From the compiler side, we need a feature flag to know when the target platform supports that bit, so we can error in cases when it's not supported (like we do with DIMs).

## Proposed API

```cs
namespace System.Runtime.CompilerServices
{
    public static partial class RuntimeFeature
    {
        /// <summary>
        /// Name of the Portable PDB feature.
        /// </summary>
        public const string PortablePdb = nameof(PortablePdb);
 
#if FEATURE_DEFAULT_INTERFACES
        /// <summary>
        /// Indicates that this version of runtime supports default interface method implementations.
        /// </summary>
        public const string DefaultImplementationsOfInterfaces = nameof(DefaultImplementationsOfInterfaces);
#endif

+       /// <summary>
+       /// Indicates that this version of runtime supports the platform default calling convention.
+       /// </summary>
+       public const string UnmanagedSignatureCallingConvention = nameof(UnmanagedSignatureCallingConvention);
 
        /// <summary>
        /// Checks whether a certain feature is supported by the Runtime.
        /// </summary>
        public static bool IsSupported(string feature)
        {
            switch (feature)
            {
                case PortablePdb:
+               case UnmanagedSignatureCallingConvention:
#if FEATURE_DEFAULT_INTERFACES
                case DefaultImplementationsOfInterfaces:
#endif
                    return true;
                case nameof(IsDynamicCodeSupported):
                    return IsDynamicCodeSupported;
                case nameof(IsDynamicCodeCompiled):
                    return IsDynamicCodeCompiled;
            }
 
            return false;
        }
    }
}
```

## Usage Examples

In the C# compiler for determining errors.

## Alternative Designs

N/A

## Risks

N/A

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Currently throws "Not Supported".

Contributes to #38133 
Fixes https://github.com/dotnet/runtime/issues/38135

/cc @lambdageek @jkotas @tannergooding @333fred @jkoritzinsky @elinor-fung @davidwrighton

### CODE REVIEW

File: src/coreclr/src/inc/corhdr.h [hunk 1]
```diff
@@ -964,7 +964,7 @@ typedef enum CorCallingConvention
     IMAGE_CEE_CS_CALLCONV_FIELD         = 0x6,
     IMAGE_CEE_CS_CALLCONV_LOCAL_SIG     = 0x7,
     IMAGE_CEE_CS_CALLCONV_PROPERTY      = 0x8,
-    IMAGE_CEE_CS_CALLCONV_UNMGD         = 0x9,
+    IMAGE_CEE_CS_CALLCONV_UNMGD         = 0x9,  // Unmanaged calling convention encoded as modopts
```
Comment (jkotas) [inline src/coreclr/src/inc/corhdr.h:967]:
```suggestion
    IMAGE_CEE_CS_CALLCONV_UNMANAGED     = 0x9,  // Unmanaged calling convention encoded as modopts
```

File: src/coreclr/src/jit/importer.cpp [hunk 2]
```diff
@@ -7560,6 +7560,11 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
         /* Get the call site sig */
         eeGetSig(pResolvedToken->token, pResolvedToken->tokenScope, pResolvedToken->tokenContext, &calliSig);
 
+        if (calliSig.callConv == CORINFO_CALLCONV_UNMANAGED)
```
Comment (jkotas) [inline src/coreclr/src/jit/importer.cpp:7563]: The VM side should be responsible for decoding the calling convention for the JIT into the a enum defined in corinfo.h (we may be able repurpose `CorInfoUnmanagedCallConv` for that).

File: src/coreclr/src/tools/Common/TypeSystem/Common/MethodDesc.cs [hunk 3]
```diff
@@ -20,6 +20,7 @@ public enum MethodSignatureFlags
         UnmanagedCallingConventionStdCall    = 0x0002,
         UnmanagedCallingConventionThisCall   = 0x0003,
         CallingConventionVarargs             = 0x0005,
+        UnmanagedCallingConventionAsModOpt   = 0x0009,
```
Comment (jkotas) [inline src/coreclr/src/tools/Common/TypeSystem/Common/MethodDesc.cs:23]:
```suggestion
        UnmanagedCallingConvention           = 0x0009,
```

File: src/coreclr/src/jit/importer.cpp (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/src/jit/importer.cpp:7563]: But that isn't how this specific API is defined. Is the suggestion we update the `CORINFO_SIG_INFO` data structure? This PR was to create the API so it can be consumed and block usage until full support can be added.

Comment (jkotas) [reply src/coreclr/src/jit/importer.cpp:7563]: > Update the `CORINFO_SIG_INFO` data structure

I think something like that will be required to get it fully supported.

> it can be consumed and block usage until full support can be added

Sounds good. Doing this in second step is fine.

File: src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/SignatureHeader.cs [hunk 4]
```diff
@@ -40,7 +40,8 @@ public SignatureCallingConvention CallingConvention
             {
                 int callingConventionOrKind = _rawValue & CallingConventionOrKindMask;
 
-                if (callingConventionOrKind > maxCallingConvention)
+                if (callingConventionOrKind > maxCallingConvention
+                    && callingConventionOrKind != (int)SignatureCallingConvention.Unmanaged)
```
Comment (jkotas) [inline src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/SignatureHeader.cs:44]: Add similar change to `SignatureKind Kind` as well?

File: src/coreclr/src/jit/importer.cpp (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/src/jit/importer.cpp:7563]: Oh wait. Shoot. I was being dense here. @BruceForstall alluded to a similar point I was missing. I can instead keep this validating in the VM side in the call to `eeGetSig()`. The VM knows that the JIT doesn't support this value so we can throw on that side and then when we add support we can then the JIT and VM in a unified fashion.

File: src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs [hunk 5]
```diff
@@ -1161,6 +1161,11 @@ private void findSig(CORINFO_MODULE_STRUCT_* module, uint sigTOK, CORINFO_CONTEX
             var methodSig = (MethodSignature)methodIL.GetObject((int)sigTOK);
             Get_CORINFO_SIG_INFO(methodSig, sig);
 
+            if (sig->callConv == CorInfoCallConv.CORINFO_CALLCONV_UNMANAGED)
+            {
+                throw new NotSupportedException();
```
Comment (tannergooding) [inline src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:1166]: Do we know what will be required to ensure it is supported?

Comment (tannergooding) [reply src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:1166]: Likewise, do we need issues for CoreRT/Mono who may want to support this as well?

File: src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/SignatureHeader.cs [hunk 6]
```diff
@@ -55,7 +56,8 @@ public SignatureKind Kind
             {
                 int callingConventionOrKind = _rawValue & CallingConventionOrKindMask;
 
-                if (callingConventionOrKind <= maxCallingConvention)
+                if (callingConventionOrKind <= maxCallingConvention
+                    && callingConventionOrKind != (int)SignatureCallingConvention.Unmanaged)
```
Comment (jkotas) [inline src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/SignatureHeader.cs:60]:
```suggestion
                    || callingConventionOrKind == (int)SignatureCallingConvention.Unmanaged)
```

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/SignatureHeader.cs:60]: Missed what this was doing.

File: src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs (hunk 5 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:1166]: @lambdageek Has been tagged on this above. This specific PR is for adding the value so it can be consumed from the Roslyn side. When it is actually implemented CrossGen2 should be okay to support this if `modopts` are automatically copied over.

File: src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/SignatureHeader.cs (hunk 6 above)
Comment (jkotas) [reply src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/SignatureHeader.cs:60]: This can use a few tests  `src\libraries\System.Reflection.Metadata\tests\Metadata\Decoding\MethodSignatureTests.cs`

File: src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs (hunk 5 above)
Comment (tannergooding) [reply src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:1166]: Do ildasm/ilasm need changes to support the new calling convention?

Comment (AaronRobinsonMSFT) [reply src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs]: That will also need support for the IL scenario.

Comment (AaronRobinsonMSFT) [reply src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs]: @BruceForstall or @briansull What is needed to add support in the IL tooling for this new MD value?

Comment (jkotas) [reply src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs]: ilasm supports arbitrary calling convention values. E.g. you can write `.method private hidebysig static callconv(9) void  foo() cil managed` today and ilasm will do the right thing for it.

ildasm is not able to disassembly it though. It errors out with `signature of method '%s' has invalid calling convention`. It would be pretty straightforward to fix that.

It would be nice to teach ilasm/ildasm to print symbolic name for this. That means adding a new line to callKind in `src\coreclr\src\ilasm\asmparse.y` and regenerating asmparse.c using internal version of yacc (it somewhere in the in the  desktop tree - related: https://github.com/dotnet/runtime/issues/4776).

Comment (AaronRobinsonMSFT) [reply src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs]: @jkotas and @tannergooding I regenerated the updated parser and added the textual grammar. I just made sure it built. I will do some testing locally to validate it works as expected.

/cc @BruceForstall @briansull

**Comment by BruceForstall** [comment]
Technically, I think we need a new JIT/EE interface GUID

**Comment by BruceForstall** [comment]
It's really worrisome to change ilasm and esp. the ilasm parser since we currently don't have automated testing (or any testing). Maybe you could figure out a way to test it by replacing the ilasm used to build the coreclr Pri-1 tests, and verify they all build (and run) without error. You'll also want to test ilasm/ildasm round tripping of the new callKind.

**Comment by AaronRobinsonMSFT** [comment]
>You'll also want to test ilasm/ildasm round tripping of the new callKind.

That seems to work for the few examples I have locally.

**Comment by jkotas** [comment]
> Technically, I think we need a new JIT/EE interface GUID

The changes in corinfo.h can be easily excluded from this change to make new JIT/EE interface GUID unnecessary (even technically).

**Comment by jkotas** [comment]
> That seems to work for the few examples I have locally.

I do not see how it can possible roundtrip with the current version of the change. I expect ildasm to error out on the `unmanaged` signatures.

**Comment by AaronRobinsonMSFT** [comment]
> > That seems to work for the few examples I have locally.
> 
> I do not see how it can possible roundtrip with the current version of the change. I expect ildasm to error out on the `unmanaged` signatures.

I compiled the following with ILAsm:

```
    IL_0020:  ldloc.3
    IL_0021:  calli      unmanaged int32(int32)
    IL_0026:  pop
    IL_0027:  ret
```

This decompiled without issue. However when I looked at it the actual IL the result is slightly off:

```
    IL_0020:  ldloc.3
    IL_0021:  calli      unmanaged cdecl int32(int32)
    IL_0026:  pop
    IL_0027:  ret
```

I observed the same thing in .NET Framework, but in that case it was slightly different because the signature token stated the correct thing (i.e. `UNMANAGED`) even though it displayed the `cdecl` above. I could be misunderstanding the overloaded "unmanaged" term here though.

It seems in this case, it isn't failing but instead falling back to some "default" value. I am still writing tests and haven't run through any actual runtime scenarios yet.

**Comment by AaronRobinsonMSFT** [comment]
@jkotas I now see what you are talking about (i.e. `PrettyPrintSig()`).

File: src/coreclr/src/ildasm/dasm.cpp [hunk 7]
```diff
@@ -3475,7 +3475,8 @@ BOOL DumpMethod(mdToken FuncToken, const char *pszClassName, DWORD dwEntryPointT
     bool bRet = FALSE;
 
     PAL_CPP_TRY {
-        if((*pComSig & IMAGE_CEE_CS_CALLCONV_MASK) > IMAGE_CEE_CS_CALLCONV_VARARG)
+        if((*pComSig & IMAGE_CEE_CS_CALLCONV_MASK) > IMAGE_CEE_CS_CALLCONV_VARARG
```
Comment (jkotas) [inline src/coreclr/src/ildasm/dasm.cpp:3478]: I would delete this error handling completely; and instead changed PrettyPrintSig to print `callconv(value)` for values that do not translated to keywords.

**Comment by AaronRobinsonMSFT** [comment]
Opinions on where some new tests should live?

Here are two I am considering:
1) `coreclr/tests/src/baseservices/callconvs`
1) `coreclr/tests/src/Interop/CallConvs`

I am leaning toward (1) since this really has nothing to do with interop but is support in the runtime for metadata related to calling conventions.

**Comment by AaronRobinsonMSFT** [comment]
I am going with (1) `coreclr/tests/src/baseservices/callconvs` above.

Well shoot. I didn't realize the IL tools were consumed from a local build. This is unfortunate. @jkotas or @tannergooding any suggestions on how to author tests here?

1) Defer tests until we can update the package
1) Hijack IL SDK to use the locally built ILAsm

The above options are what I am thinking.

**Comment by jkotas** [comment]
Add the tests as disabled, and re-enable them once we get updated tooling

**Comment by tannergooding** [comment]
I think 2 would be good. Other tools, like the C# compiler, also support overriding/dogfooding the locally built copy or a user specified copy.

**Comment by jkotas** [comment]
> Hijack IL SDK to use the locally built ILAsm

You would have an uphill battle to fight to make cross-compilation to work, and the rewire the partition dependencies. These are among the reasons why we are using LKG style model for build tools.

**Comment by AaronRobinsonMSFT** [comment]
@jkotas  @tannergooding @davidwrighton  Any additional feedback?

**Review by jkotas — APPROVED** [review]

File: src/libraries/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs [hunk 8]
```diff
@@ -2234,6 +2234,7 @@ public enum SignatureCallingConvention : byte
         ThisCall = (byte)3,
         FastCall = (byte)4,
         VarArgs = (byte)5,
+        Unmanaged = (byte)9,
```
Comment (333fred) [inline src/libraries/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs:2237]: Apologies if we've already bikeshedded this, but do we want `Unmanaged` here, or something like `UnmanagedDefault` or `PlatformDefault`? Technically 4/6 of the existing conventions are unmanaged, so just saying `Unmanaged` might be confusing?

File: src/coreclr/src/inc/corinfo.h [hunk 9]
```diff
@@ -750,7 +751,8 @@ enum CorInfoUnmanagedCallConv
     CORINFO_UNMANAGED_CALLCONV_C,
     CORINFO_UNMANAGED_CALLCONV_STDCALL,
     CORINFO_UNMANAGED_CALLCONV_THISCALL,
-    CORINFO_UNMANAGED_CALLCONV_FASTCALL
+    CORINFO_UNMANAGED_CALLCONV_FASTCALL,
```
Comment (jkotas) [inline src/coreclr/src/inc/corinfo.h:754]: Revert this change? The comment says "correspond to CorUnmanagedCallingConvention", but CorUnmanagedCallingConvention is not getting this added.

File: src/libraries/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs (hunk 8 above)
Comment (jkotas) [reply src/libraries/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs:2237]: That's a good question for API review that did not happened yet for this one.

Personally, I am fine with just `Unmanaged`. A reasonable name won't be able to capture the full semantics of this (`UnmanagedPlatformDefaultWithOptionalModOptOverridesDoNotUseForCdeclStdCallFastCall`), so it seems reasonable to just use a name that looks good.

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs:2237]: >That's a good question for API review that did not happened yet for this one.

Indeed. I will wait for Thursday API review.

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs:2237]: Handle in API review.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeFeature.cs [hunk 10]
```diff
@@ -18,6 +18,11 @@ public static partial class RuntimeFeature
         public const string DefaultImplementationsOfInterfaces = nameof(DefaultImplementationsOfInterfaces);
 #endif
 
+        /// <summary>
+        /// Indicates that this version of runtime supports the Unmanaged calling convention value.
+        /// </summary>
+        public const string UnmanagedSignatureCallingConvention = nameof(UnmanagedSignatureCallingConvention);
```
Comment (333fred) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeFeature.cs:24]: Doesn't this need to be added to the switch below?

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeFeature.cs:24]: I honestly thought the presence of the string was what was needed. Didn't realize how the system worked.

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeFeature.cs:24]: The comment "Indicates that this version of runtime supports" was the confusing part. The string constant doesn't indicate anything, it is simply the name of the feature.

@lambdageek I was going to make this the default for SPCL. Would you like me to mark this as `false` for mono until support can be added?

Comment (333fred) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeFeature.cs:24]: > I honestly thought the presence of the string was what was needed. Didn't realize how the system worked.

Well, the compiler will just look for the presence of the member. But I assumed that since everything else was in the switch, it should be consistent.

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeFeature.cs:24]: Actually the tests don't really permit making this constant dynamic so it looks like for the time being mono will indicate it is supported, but not. Unless I update the following test:

https://github.com/dotnet/runtime/blob/106cd9ebaed5f974fa8f17e153ec3e22b6576f43/src/libraries/System.Runtime/tests/System/Runtime/CompilerServices/RuntimeFeatureTests.cs#L41-L57

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeFeature.cs:24]: The current API doesn't support dynamic support so I am going to update this to lie about present mono with the expectation that at GA it will be truthful about support in mono.

/cc @lambdageek

**Comment by AaronRobinsonMSFT** [comment]
@davidwrighton Any concerns here?

**Review by davidwrighton — APPROVED** [review]
No this looks good from my pov.

File: src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/SignatureHeader.cs (hunk 4 above)
Comment (ericstj) [reply src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/SignatureHeader.cs:44]: Is there a good reason for making these values non-contiguous?

Comment (jkotas) [reply src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/SignatureHeader.cs:44]: Yes, these calling convention values share the encoding space with other signature types in the ECMA-335 file format.

Comment (ericstj) [reply src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/SignatureHeader.cs:44]: Should we make all of them part of this enum?  It seems very usual to me that we'd have non-public values of data that is stored in the metadata.

Comment (jkotas) [reply src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/SignatureHeader.cs:44]: The other values are exposed as SignatureKind enum already: https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.signaturekind?view=netcore-3.1

System.Reflection.Metadata object model uses the two different enums for the same encoding spaces as guiderails.
